dkc/core/models/folder.py

```diff
@@ -58,56 +65,19 @@ class Meta:
         'self', on_delete=models.DO_NOTHING, blank=True, null=True, related_name='+'
     )
 
-    def resolve_quota(self) -> Tuple[int, int]:
-        """
-        Resolve the quota for a given folder.
-
-        Returns the value as a tuple of the form (bytes_used, bytes_allowed).
-        """
-        folder_quota = self.root_folder.quota
-        if folder_quota.allowed is None:
-            # This indicates that this folder's quota is its owning user's quota
-            owner_quota = self.root_folder.owner.quota
-            return owner_quota.used, owner_quota.allowed
-        return folder_quota.used, folder_quota.allowed
-
-    @transaction.atomic
-    def increment_quota(self, amount: int) -> None:
-        """
-        Increments the quota(s) associated with this folder.
-
-        This function increments (or decrements, if ``amount`` is negative) the usage
-        values associated with this folder. The used amount tracked on the folder itself is
-        incremented. Additionally, if this folder has a user-assigned quota, that quota
-        usage value is also incremented. If either of these increments would exceed the max
-        allotment for this folder, a ``ValidationError`` is raised and the transaction
-        is rolled back.
-        """
-        if amount == 0:
-            return
-
-        folder_quota = self.root_folder.quota
-        folder_quota.used = models.F('used') + amount
-        folder_quota.save(update_fields=['used'])
-        folder_quota.refresh_from_db()
-
-        if folder_quota.allowed is None:
-            user_quota = self.root_folder.owner.quota
-            user_quota.used = models.F('used') + amount
-            user_quota.save(update_fields=['used'])
-            user_quota.refresh_from_db()
-            if amount > 0:
-                if user_quota.used > user_quota.allowed:
-                    raise ValidationError(
-                        'User size quota would be exceeded: '
-                        f'{user_quota.used}B > {user_quota.allowed}B.'
-                    )
-        elif amount > 0:
-            if folder_quota.used > folder_quota.allowed:
-                raise ValidationError(
-                    'Root folder size quota would be exceeded: '
-                    f'{folder_quota.used}B > {folder_quota.allowed}B.'
-                )
+    # Prevent deletion of User if it has Folders referencing it
+    owner = models.ForeignKey(User, on_delete=models.PROTECT)
+
+    # Prevent deletion of a Quota if it has Folders referencing it,
+    # unless Quota is also being deleted implicitly via a permitted CASCADE
+    quota = models.ForeignKey(Quota, null=True, on_delete=models.RESTRICT)
```

Comment:
I'd prefer not to have a nullable quota field on all folders. Only a small minority of folders will have quotas associated with them. As such, I think we should add another relation that associates (root) folders with quotas.

Reply:
I'm not sure of an elegant way to engineer this in Django's ORM which gives us the benefits of this behavior. Right now, we can seamlessly say `isinstance(root_folder.quota, Quota)` `child_folder.quota is None`. Additionally, Django certainly knows about nullable `ForeignKey` and can do predictable things when joining them, caching them, or doing a `select_related`.

A `ManyToManyField` does create an intermediate table (relation), but its engineering does not anticipate that one column might practically be unique. We'd have to add some extra constraints to ensure this, and the default API would return [a `RelatedManager`](https://docs.djangoproject.com/en/3.1/ref/models/relations/#django.db.models.fields.related.RelatedManager) on `folder.quota`. Of course, we could add even more code to wrap this under abstraction, but that would introduce more complexity, require more testing, and we might loose some hidden optimizations from Django.

The primary cost to the current design seems to be the width of a frequently-useless integer column. Note that we are still able to provide a database-level check constraint to ensure its integrity (it is nullable XOR with `parent`). Is there's another major downside to the current approach to be considered?

Reply:
I appreciate the value of being able to enforce the root-folder-only property of the quota using a check constraint. I'm just not convinced that it outweighs the cons associated with coupling these concerns together and making a messier schema. In practice, use of the Django ORM causes `select *` queries to be issued, so adding columns that we rarely care about is adding a little bit of cost for everyone on every fetch.

Ultimately though, I'm not trying to make a performance argument here. I just think we should have some default best practices when it comes to database normalization, and one of those practices is to avoid abuse of NULLs, and to try and separate concerns across relations. I like check constraints a lot, though -- and to be clear, in SQL land, we could write SQL functions that can execute arbitrary, multi-table queries and use them as check constraints, but Django doesn't currently have first-class support for that.

Reply:
> I like check constraints a lot, though -- and to be clear, in SQL land, we could write SQL functions that can execute arbitrary, multi-table queries and use them as check constraints, but Django doesn't currently have first-class support for that.

No, check constraints are expected to be immutable functions, so you can't perform subqueries in them.  https://www.postgresql.org/docs/12/ddl-constraints.html#DDL-CONSTRAINTS-CHECK-CONSTRAINTS

Reply:
Ah, lame. I guess that only works for triggers.

Reply:
> In practice, use of the Django ORM causes select * queries to be issued, so adding columns that we rarely care about is adding a little bit of cost for everyone on every fetch.

Just FYI generally, you **can** reassign the default `objects` Manager to run [`.defer`](https://docs.djangoproject.com/en/3.1/ref/models/querysets/#defer) by default on every query.